File: sdk/typescript/src/llm-providers/anthropic.ts
/**
 * SPDX-License-Identifier: Apache-2.0
 * Copyright 2026 The Plinth Authors
 *
 * Anthropic provider — wraps the official `@anthropic-ai/sdk` package.
 *
 * The provider lives behind the optional `@anthropic-ai/sdk` peer
 * dependency: `npm install @anthropic-ai/sdk`. If the package is
 * missing, {@link AnthropicProvider.create} throws
 * {@link LLMProviderNotInstalledError} with the install hint.
 *
 * The cost helper uses the published per-million-token pricing for the
 * v4.5 generation. Update {@link ANTHROPIC_PRICING} when Anthropic's
 * rates change — the table is the single source of truth for cost
 * calculation.
 */

import { LLMProviderError, LLMProviderNotInstalledError, LLMRateLimitedError } from "../errors.js";
import type { LLMMessage, LLMResponse, LLMStreamChunk } from "../types.js";
import type { LLMProvider, LLMProviderRequest } from "./types.js";

/**
 * Per-token pricing in USD. Cost = tokens * rate.
 *
 * Source: anthropic.com/pricing (v4.5 series).
 */
export const ANTHROPIC_PRICING: Record<string, { input: number; output: number }> = {
  "claude-sonnet-4-5": {
    input: 3.0 / 1_000_000,
    output: 15.0 / 1_000_000,
  },
  "claude-opus-4-5": {
    input: 15.0 / 1_000_000,
    output: 75.0 / 1_000_000,
  },
  "claude-haiku-4-5": {
    input: 0.8 / 1_000_000,
    output: 4.0 / 1_000_000,
  },
};

/**
 * Fallback pricing applied when the model name is unknown. We pick the
 * Sonnet rate because it's the most common production target — keeps
 * budgeting roughly correct on new aliases the table hasn't seen yet.
 */
const FALLBACK_PRICING = ANTHROPIC_PRICING["claude-sonnet-4-5"]!;

/** Configuration accepted by `AnthropicProvider.create`. */
export interface AnthropicProviderConfig {
  /** API key. Defaults to `process.env.ANTHROPIC_API_KEY`. */
  apiKey?: string;
  /** Override the API base URL (rare; mostly for proxies). */
  baseUrl?: string;
  /** Inject a pre-built client (mostly for tests using `vi.mock`). */
  client?: unknown;
}

/** Result of `_splitSystem` — Anthropic puts the system prompt outside `messages`. */
interface SplitMessages {
  system: string | null;
  messages: Array<{ role: "user" | "assistant"; content: string }>;
}

/**
 * Plinth's adapter over `@anthropic-ai/sdk`.
 *
 * Construction is async because `@anthropic-ai/sdk` is loaded via
 * dynamic import — this keeps the SDK tree-shakeable and lets users
 * who never call `useProvider("anthropic")` skip the install.
 */
export class AnthropicProvider implements LLMProvider {
  readonly name = "anthropic";

  /**
   * Build the provider, lazily importing `@anthropic-ai/sdk`.
   *
   * Throws {@link LLMProviderNotInstalledError} if the package isn't
   * installed; throws {@link LLMProviderError} with no `statusCode` if
   * the constructor itself rejects (e.g. malformed API key).
   */
  static async create(config: AnthropicProviderConfig = {}): Promise<AnthropicProvider> {
    if (config.client !== undefined) {
      // Test path — caller supplies a fake `Anthropic` client object.
      return new AnthropicProvider(config.client);
    }
    let mod: { default?: unknown; Anthropic?: unknown };
    try {
      // The vendor SDK is an optional peer dep; the import below is
      // only valid when the user has installed it. We use a runtime
      // string to suppress TS's static module resolution, then cast.
      const importPath = "@anthropic-ai/sdk";
      mod = (await import(importPath)) as {
        default?: unknown;
        Anthropic?: unknown;
      };
    } catch (err) {
      throw new LLMProviderNotInstalledError(
        "The 'anthropic' provider requires @anthropic-ai/sdk. " +
          "Install it with: npm install @anthropic-ai/sdk",
        { cause: (err as Error).message },
      );
    }
    // The SDK's surface has shifted between major versions — the
    // constructor lives at `default` (v0.28+) or `Anthropic` (older).
    const Ctor = (mod.default ?? mod.Anthropic) as
      | (new (opts: Record<string, unknown>) => unknown)
      | undefined;
    if (typeof Ctor !== "function") {
      throw new LLMProviderNotInstalledError(
        "The installed @anthropic-ai/sdk does not export the expected " +
          "constructor. Upgrade to >=0.30.0.",
      );
    }
    const apiKey = config.apiKey ?? process.env.ANTHROPIC_API_KEY;
    const opts: Record<string, unknown> = {};
    if (apiKey) opts.apiKey = apiKey;
    if (config.baseUrl) opts.baseURL = config.baseUrl;
    const client = new Ctor(opts);
    return new AnthropicProvider(client);
  }

  /**
   * Direct constructor — caller supplies a configured client.
   *
   * `LLMClient.useProvider("anthropic", ...)` calls
   * {@link AnthropicProvider.create} and awaits the result; tests can
   * `new AnthropicProvider(fakeClient)` directly.
   */
  constructor(private readonly client: unknown) {}

  // ------------------------------------------------------------------
  // Message translation
  // ------------------------------------------------------------------

  /**
   * Split system messages out and reformat the rest.
   *
   * Anthropic's API takes the system prompt as a top-level argument;
   * we concatenate multiple system messages with a blank line so callers
   * can compose them without rebuilding the prompt themselves.
   */
  private static splitSystem(messages: LLMMessage[]): SplitMessages {
    const systemParts: string[] = [];
    const chat: Array<{ role: "user" | "assistant"; content: string }> = [];
    for (const msg of messages) {
      if (msg.role === "system") {
        if (msg.content) systemParts.push(msg.content);
        continue;
      }
      // Anthropic's role enum is `user` | `assistant`. Tool messages
      // aren't supported on the simple-chat surface; collapse anything
      // unknown to `user` so the call doesn't 400.
      const apiRole: "user" | "assistant" = msg.role === "assistant" ? "assistant" : "user";
      chat.push({ role: apiRole, content: msg.content });
    }
    return {
      system: systemParts.length > 0 ? systemParts.join("\n\n") : null,
      messages: chat,
    };
  }

  /** Pull the assistant's text out of an Anthropic `Message`. */
  private static extractContent(message: unknown): string {
    const blocks = (message as { content?: unknown }).content;
    if (!Array.isArray(blocks)) return "";
    const parts: string[] = [];
    for (const block of blocks) {
      if (block && typeof block === "object" && typeof (block as { text?: unknown }).text === "string") {
        parts.push((block as { text: string }).text);
      }
    }
    return parts.join("");
  }

  /** Return `[input, output]` token counts from a response, defaulting to 0. */
  private static extractUsage(message: unknown): [number, number] {
    const usage = (message as { usage?: unknown }).usage;
    if (!usage || typeof usage !== "object") return [0, 0];
    const u = usage as Record<string, unknown>;
    const inT = Number(u.input_tokens ?? u.prompt_tokens ?? 0);
    const outT = Number(u.output_tokens ?? u.completion_tokens ?? 0);
    return [
      Number.isFinite(inT) ? inT : 0,
      Number.isFinite(outT) ? outT : 0,
    ];
  }

  /** Best-effort conversion to a plain object for `raw`. */
  private static toDict(message: unknown): Record<string, unknown> {
    if (message && typeof message === "object") {
      try {
        return JSON.parse(JSON.stringify(message)) as Record<string, unknown>;
      } catch {
        return { raw: String(message) };
      }
    }
    return { raw: String(message) };
  }

  // ------------------------------------------------------------------
  // Cost
  // ------------------------------------------------------------------

  estimateCostUsd(model: string, inputTokens: number, outputTokens: number): number {
    const pricing = ANTHROPIC_PRICING[model] ?? FALLBACK_PRICING;
    return inputTokens * pricing.input + outputTokens * pricing.output;
  }

  // ------------------------------------------------------------------
  // Error mapping
  // ------------------------------------------------------------------

  /** Translate a vendor SDK error to a Plinth `LLMProviderError`. */
  private wrapError(exc: unknown): Error {
    // The SDK's typed errors carry `status` and `headers`. We don't
    // `instanceof` the classes — that would require a static import
    // and defeat the lazy-import design. Instead we duck-type on the
    // shape: an `APIStatusError` has `status` (number) and may have
    // `headers` (object). Rate-limit specifically also has status 429.
    const e = exc as {
      status?: number;
      message?: string;
      response?: { headers?: Record<string, string>; status?: number };
      body?: unknown;
      headers?: Record<string, string>;
      name?: string;
    };
    const message = typeof e.message === "string" ? e.message : String(exc);
    const status = e.status ?? e.response?.status;
    if (status === 429 || (typeof e.name === "string" && /rate.?limit/i.test(e.name))) {
      const retryAfter = parseRetryAfter(e);
      return new LLMRateLimitedError(message, {
        retryAfterSeconds: retryAfter,
        statusCode: status ?? 429,
        body: e.body,
        provider: this.name,
      });
    }
    if (typeof status === "number") {
      return new LLMProviderError(message, {
        statusCode: status,
        body: e.body,
        provider: this.name,
      });
    }
    // Connection / unknown error.
    return new LLMProviderError(message, { provider: this.name });
  }

  // ------------------------------------------------------------------
  // Surface
  // ------------------------------------------------------------------

  async complete(req: LLMProviderRequest): Promise<LLMResponse> {
    const split = AnthropicProvider.splitSystem(req.messages);
    const apiArgs: Record<string, unknown> = {
      model: req.model,
      messages: split.messages,
      // Anthropic requires `max_tokens` — default 1024 matches Python.
      max_tokens: req.maxTokens ?? 1024,
    };
    if (split.system !== null) apiArgs.system = split.system;
    if (req.temperature !== undefined) apiArgs.temperature = req.temperature;
    if (req.topP !== undefined) apiArgs.top_p = req.topP;
    if (req.stopSequences !== undefined && req.stopSequences.length > 0) {
      apiArgs.stop_sequences = req.stopSequences;
    }
    if (req.extra) Object.assign(apiArgs, req.extra);

    let message: unknown;
    try {
      const messages = (this.client as { messages?: unknown }).messages as
        | { create?: (args: unknown) => Promise<unknown> }
        | undefined;
      if (!messages || typeof messages.create !== "function") {
        throw new Error("Anthropic client is missing `messages.create`");
      }
      message = await messages.create(apiArgs);
    } catch (err) {
      throw this.wrapError(err);
    }

    const text = AnthropicProvider.extractContent(message);
    const [inT, outT] = AnthropicProvider.extractUsage(message);
    const stopReason = (message as { stop_reason?: unknown }).stop_reason;
    const finishReason = typeof stopReason === "string" && stopReason ? stopReason : "stop";
    const reportedModel = (message as { model?: unknown }).model;
    return {
      content: text,
      model: typeof reportedModel === "string" ? reportedModel : req.model,
      finishReason,
      inputTokens: inT,
      outputTokens: outT,
      costUsd: this.estimateCostUsd(req.model, inT, outT),
      durationMs: 0, // populated by `LLMClient` using its own clock
      provider: this.name,
      raw: AnthropicProvider.toDict(message),
    };
  }

  async *stream(req: LLMProviderRequest): AsyncGenerator<LLMStreamChunk, void, void> {
    const split = AnthropicProvider.splitSystem(req.messages);
    const apiArgs: Record<string, unknown> = {
      model: req.model,
      messages: split.messages,
      max_tokens: req.maxTokens ?? 1024,
      stream: true,
    };
    if (split.system !== null) apiArgs.system = split.system;
    if (req.temperature !== undefined) apiArgs.temperature = req.temperature;
    if (req.topP !== undefined) apiArgs.top_p = req.topP;
    if (req.stopSequences !== undefined && req.stopSequences.length > 0) {
      apiArgs.stop_sequences = req.stopSequences;
    }
    if (req.extra) Object.assign(apiArgs, req.extra);

    let stream: AsyncIterable<unknown>;
    try {
      const messages = (this.client as { messages?: unknown }).messages as
        | {
            create?: (args: unknown) => Promise<AsyncIterable<unknown>>;
            stream?: (args: unknown) => AsyncIterable<unknown>;
          }
        | undefined;
      if (!messages) throw new Error("Anthropic client is missing `messages`");
      // Prefer `messages.stream` (returns AsyncIterable directly) when
      // present; fall back to `messages.create` with `stream: true`.
      if (typeof messages.stream === "function") {
        stream = messages.stream(apiArgs);
      } else if (typeof messages.create === "function") {
        stream = await messages.create(apiArgs);
      } else {
        throw new Error("Anthropic client is missing both `messages.stream` and `messages.create`");
      }
    } catch (err) {
      throw this.wrapError(err);
    }

    let finishReason: string | undefined;
    let lastRaw: Record<string, unknown> = {};
    try {
      for await (const event of stream) {
        // The SDK emits typed events; we care about
        //   `content_block_delta` (text deltas)
        //   `message_delta` (carries stop_reason)
        //   `message_stop` (terminal marker)
        if (event && typeof event === "object") {
          const e = event as Record<string, unknown>;
          const type = typeof e.type === "string" ? e.type : "";
          if (type === "content_block_delta") {
            const delta = e.delta as { type?: unknown; text?: unknown } | undefined;
            if (delta && typeof delta.text === "string" && delta.text.length > 0) {
              yield { delta: delta.text, raw: {} };
            }
          } else if (type === "message_delta") {
            const inner = (e.delta as { stop_reason?: unknown }) ?? {};
            if (typeof inner.stop_reason === "string") finishReason = inner.stop_reason;
          } else if (typeof (e as { delta?: unknown }).delta === "string") {
            // Some test doubles emit `{ delta: "abc" }` directly.
            yield { delta: (e as { delta: string }).delta, raw: {} };
          }
          lastRaw = e as Record<string, unknown>;
        }
      }
    } catch (err) {
      throw this.wrapError(err);
    }

    yield { delta: "", finishReason: finishReason ?? "stop", raw: lastRaw };
  }
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/**
 * Extract a `retry-after` hint from an SDK error.
 *
 * Anthropic surfaces the header on `error.headers` (newer SDKs) or
 * `error.response.headers`. A best-effort lookup keeps the SDK working
 * when the upstream layout shifts between minor versions.
 */
function parseRetryAfter(err: {
  headers?: Record<string, string>;
  response?: { headers?: Record<string, string> };
}): number | null {
  const headers = err.headers ?? err.response?.headers;
  if (!headers) return null;
  const raw = headers["retry-after"] ?? headers["Retry-After"] ?? null;
  if (raw === null) return null;
  const n = Number(raw);
  return Number.isFinite(n) ? n : null;
}

File: sdk/python/src/plinth/llm_providers/anthropic.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 The Plinth Authors
"""Anthropic provider — wraps the official ``anthropic`` SDK.

The provider lives behind the optional ``[anthropic]`` extra:
``pip install 'plinth[anthropic]'``. If the package is missing the
import will fail and :func:`plinth.llm_providers.build_provider`
re-raises as :class:`~plinth.exceptions.LLMProviderNotInstalled` with
the install hint.

The cost helper uses the published per-million-token pricing for the
v4.5 generation. Exact figures are taken from Anthropic's pricing page.
Update :data:`ANTHROPIC_PRICING` when the published rates change — the
table is the single source of truth for cost calculation.
"""

from __future__ import annotations

import os
from collections.abc import AsyncIterator, Iterator
from typing import Any

from ..exceptions import (
    LLMProviderError,
    LLMRateLimited,
)
from ..models import LLMMessage, LLMResponse, LLMStreamChunk

# ``anthropic`` is an optional dependency. The import sits at module load
# so :func:`build_provider` can catch :class:`ImportError` and raise the
# typed :class:`LLMProviderNotInstalled`. Inside the provider we also
# import ``anthropic.APIStatusError`` for status-code mapping; that
# import is local to the call sites because it lives behind the same
# extra.
import anthropic  # type: ignore[import-untyped]  # noqa: E402


#: Per-token pricing in USD. Numbers are USD/token (not USD/Mtok), so
#: cost = tokens * rate.
#:
#: Source: anthropic.com/pricing (v4.5 series).
ANTHROPIC_PRICING: dict[str, dict[str, float]] = {
    "claude-sonnet-4-5": {
        "input": 3.00 / 1_000_000,
        "output": 15.00 / 1_000_000,
    },
    "claude-opus-4-5": {
        "input": 15.00 / 1_000_000,
        "output": 75.00 / 1_000_000,
    },
    "claude-haiku-4-5": {
        "input": 0.80 / 1_000_000,
        "output": 4.00 / 1_000_000,
    },
}

#: Fallback pricing applied when the model name is unknown. Picks the
#: Sonnet rate because it's the most common production target — keeps
#: budgeting roughly correct on new aliases the table hasn't seen yet.
_FALLBACK_PRICING: dict[str, float] = ANTHROPIC_PRICING["claude-sonnet-4-5"]


class AnthropicProvider:
    """Plinth's adapter over ``anthropic.Anthropic``.

    Args:
        api_key: Anthropic API key. Defaults to ``ANTHROPIC_API_KEY`` from
            the environment when omitted.
        base_url: Override the API base URL (rare; mostly for proxies).
        client_options: Extra keyword args forwarded to the underlying
            ``anthropic.Anthropic`` constructor — useful for custom
            HTTP clients in tests.
    """

    name = "anthropic"

    def __init__(
        self,
        *,
        api_key: str | None = None,
        base_url: str | None = None,
        client_options: dict[str, Any] | None = None,
    ) -> None:
        resolved_key = api_key or os.environ.get("ANTHROPIC_API_KEY")
        kwargs: dict[str, Any] = dict(client_options or {})
        if resolved_key:
            kwargs.setdefault("api_key", resolved_key)
        if base_url:
            kwargs.setdefault("base_url", base_url)
        self._client = anthropic.Anthropic(**kwargs)
        # Async client lazily constructed: ``anthropic.AsyncAnthropic``
        # opens its own connection pool and we don't want to pay for it
        # in sync-only programs.
        self._async_client: anthropic.AsyncAnthropic | None = None
        self._async_kwargs = kwargs

    def _get_async(self) -> anthropic.AsyncAnthropic:
        if self._async_client is None:
            self._async_client = anthropic.AsyncAnthropic(**self._async_kwargs)
        return self._async_client

    # ------------------------------------------------------------------
    # Message translation
    # ------------------------------------------------------------------

    @staticmethod
    def _split_system(
        messages: list[LLMMessage] | list[dict[str, Any]],
    ) -> tuple[str | None, list[dict[str, Any]]]:
        """Split system messages out and reformat the rest.

        Anthropic's API takes the system prompt as a top-level argument
        and the rest as a ``messages`` list of ``{role, content}``. We
        concatenate multiple system messages with a blank line so callers
        can compose them without rebuilding the prompt themselves.
        """
        system_parts: list[str] = []
        chat: list[dict[str, Any]] = []
        for msg in messages:
            role = msg.role if isinstance(msg, LLMMessage) else msg.get("role")
            content = (
                msg.content if isinstance(msg, LLMMessage) else msg.get("content", "")
            )
            if role == "system":
                if content:
                    system_parts.append(content)
                continue
            # Anthropic's role enum is ``user`` | ``assistant``. Tool
            # messages aren't supported on the message-level surface in
            # the simple-chat API path; we collapse any unknown role to
            # ``user`` so the call doesn't 400.
            api_role = role if role in ("user", "assistant") else "user"
            chat.append({"role": api_role, "content": content})
        system = "\n\n".join(system_parts) if system_parts else None
        return system, chat

    @staticmethod
    def _extract_content(message: Any) -> str:
        """Pull the assistant's text out of an Anthropic ``Message``.

        Anthropic returns ``content`` as a list of blocks; for plain
        text completions there's a single ``TextBlock``. We concatenate
        every text block and ignore other block types (which the simple
        completion path shouldn't produce, but defensively…).
        """
        content_blocks = getattr(message, "content", None) or []
        parts: list[str] = []
        for block in content_blocks:
            text = getattr(block, "text", None)
            if isinstance(text, str):
                parts.append(text)
            elif isinstance(block, dict) and isinstance(block.get("text"), str):
                parts.append(block["text"])
        return "".join(parts)

    @staticmethod
    def _extract_usage(message: Any) -> tuple[int, int]:
        """Return ``(input_tokens, output_tokens)`` from a response."""
        usage = getattr(message, "usage", None)
        if usage is None:
            return 0, 0
        in_tok = (
            getattr(usage, "input_tokens", None) or getattr(usage, "prompt_tokens", 0)
        )
        out_tok = (
            getattr(usage, "output_tokens", None)
            or getattr(usage, "completion_tokens", 0)
        )
        return int(in_tok or 0), int(out_tok or 0)

    @staticmethod
    def _to_dict(message: Any) -> dict[str, Any]:
        """Best-effort conversion to a plain dict for ``raw``."""
        if hasattr(message, "model_dump"):
            try:
                return dict(message.model_dump())
            except Exception:  # noqa: BLE001 - best-effort serialization
                pass
        if hasattr(message, "dict"):
            try:
                return dict(message.dict())  # type: ignore[call-arg]
            except Exception:  # noqa: BLE001
                pass
        return {"raw": str(message)}

    # ------------------------------------------------------------------
    # Cost
    # ------------------------------------------------------------------

    def estimate_cost_usd(
        self, model: str, input_tokens: int, output_tokens: int
    ) -> float:
        """Compute USD cost using :data:`ANTHROPIC_PRICING`."""
        pricing = ANTHROPIC_PRICING.get(model, _FALLBACK_PRICING)
        return input_tokens * pricing["input"] + output_tokens * pricing["output"]

    # ------------------------------------------------------------------
    # Error mapping
    # ------------------------------------------------------------------

    def _wrap_error(self, exc: BaseException) -> Exception:
        """Translate an SDK error to a Plinth :class:`LLMProviderError`."""
        # ``anthropic.RateLimitError`` is a 429 subclass on
        # ``APIStatusError``. Treat it specially so the retry loop knows
        # to honour ``Retry-After``.
        if isinstance(exc, anthropic.RateLimitError):
            retry_after = _parse_retry_after(exc)
            return LLMRateLimited(
                str(exc),
                retry_after_seconds=retry_after,
                status_code=getattr(exc, "status_code", 429),
                body=_response_body(exc),
                provider=self.name,
            )
        if isinstance(exc, anthropic.APIStatusError):
            return LLMProviderError(
                str(exc),
                status_code=getattr(exc, "status_code", None),
                body=_response_body(exc),
                provider=self.name,
            )
        if isinstance(exc, anthropic.APIConnectionError):
            return LLMProviderError(
                f"connection error: {exc}",
                status_code=None,
                provider=self.name,
            )
        # Unknown — wrap so downstream code can ``except LLMError``.
        return LLMProviderError(str(exc), provider=self.name)

    # ------------------------------------------------------------------
    # Sync surface
    # ------------------------------------------------------------------

    def complete(
        self,
        *,
        model: str,
        messages: list[LLMMessage] | list[dict[str, Any]],
        max_tokens: int = 1024,
        temperature: float | None = None,
        **kwargs: Any,
    ) -> LLMResponse:
        """One-shot completion. Translates errors via :meth:`_wrap_error`."""
        system, chat = self._split_system(messages)
        api_kwargs: dict[str, Any] = {
            "model": model,
            "messages": chat,
            "max_tokens": max_tokens,
        }
        if system is not None:
            api_kwargs["system"] = system
        if temperature is not None:
            api_kwargs["temperature"] = temperature
        # Pass-through extras the caller wants (top_p, stop_sequences …).
        api_kwargs.update(kwargs)

        try:
            message = self._client.messages.create(**api_kwargs)
        except Exception as exc:  # noqa: BLE001
            raise self._wrap_error(exc) from exc

        text = self._extract_content(message)
        in_tok, out_tok = self._extract_usage(message)
        finish_reason = getattr(message, "stop_reason", None) or "stop"
        return LLMResponse(
            content=text,
            model=getattr(message, "model", model),
            finish_reason=str(finish_reason),
            input_tokens=in_tok,
            output_tokens=out_tok,
            cost_usd=self.estimate_cost_usd(model, in_tok, out_tok),
            duration_ms=0,  # populated by LLMClient using its own clock
            provider=self.name,
            raw=self._to_dict(message),
        )

    def stream(
        self,
        *,
        model: str,
        messages: list[LLMMessage] | list[dict[str, Any]],
        max_tokens: int = 1024,
        temperature: float | None = None,
        **kwargs: Any,
    ) -> Iterator[LLMStreamChunk]:
        """Yield incremental chunks from the Anthropic streaming API."""
        system, chat = self._split_system(messages)
        api_kwargs: dict[str, Any] = {
            "model": model,
            "messages": chat,
            "max_tokens": max_tokens,
        }
        if system is not None:
            api_kwargs["system"] = system
        if temperature is not None:
            api_kwargs["temperature"] = temperature
        api_kwargs.update(kwargs)

        try:
            with self._client.messages.stream(**api_kwargs) as stream:
                for text in stream.text_stream:
                    yield LLMStreamChunk(delta=text)
                final = stream.get_final_message()
                yield LLMStreamChunk(
                    delta="",
                    finish_reason=str(getattr(final, "stop_reason", "stop") or "stop"),
                    raw=self._to_dict(final),
                )
        except Exception as exc:  # noqa: BLE001
            raise self._wrap_error(exc) from exc

    # ------------------------------------------------------------------
    # Async surface
    # ------------------------------------------------------------------

    async def acomplete(
        self,
        *,
        model: str,
        messages: list[LLMMessage] | list[dict[str, Any]],
        max_tokens: int = 1024,
        temperature: float | None = None,
        **kwargs: Any,
    ) -> LLMResponse:
        """Async equivalent of :meth:`complete`."""
        client = self._get_async()
        system, chat = self._split_system(messages)
        api_kwargs: dict[str, Any] = {
            "model": model,
            "messages": chat,
            "max_tokens": max_tokens,
        }
        if system is not None:
            api_kwargs["system"] = system
        if temperature is not None:
            api_kwargs["temperature"] = temperature
        api_kwargs.update(kwargs)

        try:
            message = await client.messages.create(**api_kwargs)
        except Exception as exc:  # noqa: BLE001
            raise self._wrap_error(exc) from exc

        text = self._extract_content(message)
        in_tok, out_tok = self._extract_usage(message)
        finish_reason = getattr(message, "stop_reason", None) or "stop"
        return LLMResponse(
            content=text,
            model=getattr(message, "model", model),
            finish_reason=str(finish_reason),
            input_tokens=in_tok,
            output_tokens=out_tok,
            cost_usd=self.estimate_cost_usd(model, in_tok, out_tok),
            duration_ms=0,
            provider=self.name,
            raw=self._to_dict(message),
        )

    async def astream(
        self,
        *,
        model: str,
        messages: list[LLMMessage] | list[dict[str, Any]],
        max_tokens: int = 1024,
        temperature: float | None = None,
        **kwargs: Any,
    ) -> AsyncIterator[LLMStreamChunk]:
        """Async streaming counterpart of :meth:`stream`."""
        client = self._get_async()
        system, chat = self._split_system(messages)
        api_kwargs: dict[str, Any] = {
            "model": model,
            "messages": chat,
            "max_tokens": max_tokens,
        }
        if system is not None:
            api_kwargs["system"] = system
        if temperature is not None:
            api_kwargs["temperature"] = temperature
        api_kwargs.update(kwargs)

        try:
            async with client.messages.stream(**api_kwargs) as stream:
                async for text in stream.text_stream:
                    yield LLMStreamChunk(delta=text)
                final = await stream.get_final_message()
                yield LLMStreamChunk(
                    delta="",
                    finish_reason=str(getattr(final, "stop_reason", "stop") or "stop"),
                    raw=self._to_dict(final),
                )
        except Exception as exc:  # noqa: BLE001
            raise self._wrap_error(exc) from exc


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def _parse_retry_after(exc: BaseException) -> float | None:
    """Extract a ``retry-after`` hint from an SDK error, if present.

    Anthropic surfaces the header on ``exc.response`` (when available).
    A best-effort lookup keeps the SDK working when the upstream layout
    shifts between minor versions.
    """
    response = getattr(exc, "response", None)
    if response is None:
        return None
    headers = getattr(response, "headers", None) or {}
    raw = headers.get("retry-after") or headers.get("Retry-After")
    if raw is None:
        return None
    try:
        return float(raw)
    except (TypeError, ValueError):
        return None


def _response_body(exc: BaseException) -> Any:
    """Return the raw body text of the SDK error, when available."""
    body = getattr(exc, "body", None)
    if body is not None:
        return body
    response = getattr(exc, "response", None)
    if response is not None:
        try:
            return response.text
        except Exception:  # noqa: BLE001
            return None
    return None


__all__ = ["ANTHROPIC_PRICING", "AnthropicProvider"]
